lib/index.ts

```diff
@@ -0,0 +1,295 @@
+    });
+  }
+
+  static fromEntries<T extends BaseEntry>(entries: T[], options: FSTree.StaticOptions = {}) {
+    return new this({
+      entries: entries,
+      sortAndExpand: options.sortAndExpand,
+    });
+  }
+
+  get size() {
+    return this.entries.length;
+  }
+
+  addEntries(entries: T[], options?: FSTree.StaticOptions) {
+    if (!Array.isArray(entries)) {
+      throw new TypeError('entries must be an array');
+    }
+    if (options !== null && typeof options === 'object' && options.sortAndExpand) {
+      sortAndExpand(entries);
+    } else {
+      validateSortedUnique(entries);
+    }
+    let fromIndex = 0;
+    let toIndex = 0;
+    while (fromIndex < entries.length) {
+      while (toIndex < this.entries.length &&
+        this.entries[toIndex].relativePath < entries[fromIndex].relativePath) {
+        toIndex++;
+      }
+      if (toIndex < this.entries.length &&
+        this.entries[toIndex].relativePath === entries[fromIndex].relativePath) {
+        this.entries.splice(toIndex, 1, entries[fromIndex++]);
+      } else {
+        this.entries.splice(toIndex++, 0, entries[fromIndex++]);
+      }
+    }
+  }
+
+  addPaths(paths: string[], options?: FSTree.StaticOptions) {
+    const entries = paths.map(path => {
+      // TODO:
+      // addPths + a custom isEqual comparator + custom Entry types are actually incompatible
+      // As a addPaths will not abide by the custom Entry type
+      // and will make this.entries be contain a mixture of types.
+      // isEqual's arguments will then be typed incorrectly
+      //
+      // We should likely just deprecate `addPaths` in-favor of addEntries,
+      // which correctly externalizes the creation of entry
+      return new Entry(path, 0, ARBITRARY_START_OF_TIME) as T;
+    });
+
+    this.addEntries(entries, options);
+  }
+
+  forEach(fn: (entry: T, index: number, collection: T[]) => void, context: any) {
+    this.entries.forEach(fn, context);
+  }
+
+  calculatePatch<K extends BaseEntry>(theirFSTree: FSTree<K>, isEqual?: (a: T, b: K) => boolean): FSTree.Patch {
+    if (arguments.length > 1 && typeof isEqual !== 'function') {
+      throw new TypeError('calculatePatch\'s second argument must be a function');
+    }
+
+    if (typeof isEqual !== 'function') {
+      isEqual = (this.constructor as typeof FSTree).defaultIsEqual;
+    }
+
+    const ours = this.entries;
+    const theirs = theirFSTree.entries;
+    const additions: FSTree.Patch = [];
+    const removals: FSTree.Patch = [];
+
+    let i = 0;
+    let j = 0;
+
+    let command;
+
+    while (i < ours.length && j < theirs.length) {
+      let x = ours[i];
+      let y = theirs[j];
+
+      if (x.relativePath < y.relativePath) {
+        // ours
+        i++;
+
+        removals.push(removeOperation(x));
+
+        // remove additions
+      } else if (x.relativePath > y.relativePath) {
+        // theirs
+        j++;
+        additions.push(addOperation(y));
+      } else {
+        if (!isEqual(x, y)) {
```

Comment:
I'm actually not super sure how Typescript thinks this is valid yet. I'll have to do more investigation.

Basically, if T,K do not extend from or implement DefaultEntry, and isEqual is the defaultIsEqual. Then I think we should have a type mismatch...

Reply:
Because defaultIsEqual satisfies the a: T, b: K but it requires a: K, b: K.